File: R/gamma.R
#' Sample from gamma distribution with shape `shape` and scale `scale`.
#'
#' @param n Number of samples to draw.
#' @param shape Shape parameter of the gamma distribution.
#' @param scale Scale parameter of the gamma distribution.
#' @param random_seed Random seed for reproducibility.
#' @param method Underlying C++ sampling method to use ("std" or "boost").
#'
#' @return A vector of sampled values of length `n`.
#' @export
#'
#' @examples
#' draws <- sample_gamma(n = 10, shape = 1, scale = 1, random_seed = 1, method = "std")
sample_gamma <- function(n, shape = 1, scale = 1, random_seed = -1, method = "std") {
  if (method == "std") {
    return(sample_gamma_std_cpp(n, shape, scale, random_seed))
  } else if (method == "boost") {
    return(sample_gamma_boost_cpp(n, shape, scale, random_seed))
  } else {
    stop("Unknown method")
  }
}

File: src/gamma.cpp
#include <cpp11.hpp>
#include <random>
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/gamma_distribution.hpp>
#include <distributions.h>
#include "utils.h"

[[cpp11::register]]
cpp11::writable::doubles sample_gamma_std_cpp(int n, double shape, double scale, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    std::mt19937 gen = create_std_mt19937(random_seed);

    // Initialize and sample from distribution
    std::gamma_distribution<double> gamma_dist(shape, scale);
    for (int i = 0; i < n; i++) {
      output[i] = gamma_dist(gen);
    }

    return output;
}

[[cpp11::register]]
cpp11::writable::doubles sample_gamma_boost_cpp(int n, double shape, double scale, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    boost::random::mt19937 gen = create_boost_mt19937(random_seed);

    // Initialize and sample from distribution
    boost::random::gamma_distribution<double> gamma_dist(shape, scale);
    for (int i = 0; i < n; i++) {
      output[i] = gamma_dist(gen);
    }

    return output;
}

[[cpp11::register]]
cpp11::writable::doubles sample_gamma_custom_cpp(int n, double shape, double scale, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    std::mt19937 gen = create_std_mt19937(random_seed);

    // Sample from the gamma distribution
    for (int i = 0; i < n; i++) {
      output[i] = sample_gamma(gen, shape, scale);
    }

    return output;
}
